src/cpp/statistics/rtps/writer/StatisticsWriterImpl.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file StatisticsWriterImpl.hpp
+ */
+
+#include <fastdds/rtps/writer/RTPSWriter.h>
+#include <statistics/rtps/StatisticsBase.hpp>
+
+using namespace eprosima::fastdds::statistics;
+
+using eprosima::fastrtps::RecursiveTimedMutex;
+using eprosima::fastrtps::rtps::RTPSWriter;
+using eprosima::fastrtps::rtps::GUID_t;
+
+StatisticsWriterImpl::StatisticsWriterImpl()
+{
+    init_statistics<StatisticsWriterAncillary>();
+}
+
+StatisticsWriterAncillary* StatisticsWriterImpl::get_members() const
+{
+    static_assert(
+        std::is_base_of<StatisticsAncillary, StatisticsWriterAncillary>::value,
+        "Auxiliary structure must derive from StatisticsAncillary");
+
+    return static_cast<StatisticsWriterAncillary*>(get_aux_members());
+}
+
+RecursiveTimedMutex& StatisticsWriterImpl::get_statistics_mutex()
+{
+    static_assert(
+        std::is_base_of<StatisticsWriterImpl, RTPSWriter>::value,
+        "Must be call from a writer.");
+
+    return static_cast<RTPSWriter*>(this)->getMutex();
+}
+
+const GUID_t& StatisticsWriterImpl::get_guid()
+{
+    using eprosima::fastrtps::rtps::RTPSWriter;
+
+    static_assert(
+        std::is_base_of<StatisticsWriterImpl, RTPSWriter>::value,
+        "This method should be called from an actual RTPSWriter");
+
+    return static_cast<RTPSWriter*>(this)->getGuid();
+}
+
+void StatisticsWriterImpl::on_data()
+{
+    EntityCount notification;
+    notification.guid(to_statistics_type(get_guid()));
+
+    {
+        std::lock_guard<fastrtps::RecursiveTimedMutex> lock(get_statistics_mutex());
+        notification.count(++get_members()->data_counter);
+    }
+
+    // Callback
+    Data d;
+    // note that the setter sets RESENT_DATAS by default
+    d.entity_count(notification);
+    d._d(EventKind::DATA_COUNT);
+
+    for_each_listener([&d](const std::shared_ptr<IListener>& l)
+            {
+                l->on_statistics_data(d);
+            });
+}
+
+void StatisticsWriterImpl::on_data_frag()
+{
+    // there is no specific EventKind thus it will be redirected to DATA_COUNT
+    on_data();
+}
+
+void StatisticsWriterImpl::on_heartbeat(
+        uint32_t count)
+{
+    EntityCount notification;
+    notification.guid(to_statistics_type(get_guid()));
+    notification.count(count);
+
+    // Callback
+    Data d;
+    // note that the setter sets RESENT_DATAS by default
```

Comment:
Is this comment updated?

Reply:
Highlights that the following code is performing the actual callbacks. Let's use something more descriptive like *Perform the callbacks*

Reply:
I was referring to this other comment (the one just before the conversation):

```
    // note that the setter sets RESENT_DATAS by default
```

In this case we are notifying HEARTBETS and not RESENT_DATAS. I am not sure what this comment is explaining.

Reply:
Ok, it's a little tricky.
It has to do with the *Fast-DDS-Gen* generated code. Basically in *include\fastdds\statistics\types.idl* is defined `Data` as a union of several types using *EventKind* as discriminant. But a specific type like `EntityCount` shares several values of the enumeration.
```idl
union Data switch(EventKind)
{
...
    case RESENT_DATAS:
    case HEARTBEAT_COUNT:
    case ACKNACK_COUNT:
    case NACKFRAG_COUNT:
    case GAP_COUNT:
    case DATA_COUNT:
    case PDP_PACKETS:
    case EDP_PACKETS:
        EntityCount entity_count;
....
};
```
The generated C++ code setters only take as argument the type (`EntityCount`) and ignore the discriminant value, thus, by default all are `RESENT_DATAS` and the real value must be specificied later using `Data::_d(...)`.
There is a backlog task associated [11099]